File: nvpl_blas/c/zherk.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_zherk
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t K = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldc = 2;
    double alpha = 1.0f;
    double beta = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_TRANSPOSE trans = CblasNoTrans;
    nvpl_dcomplex_t * A = NULL;
    nvpl_dcomplex_t * C = NULL;

    printf("\nExample: cblas_zherk for the Hermitian rank-k update\n\n");
    printf("#### args: n=%" PRId64 ", k=%" PRId64 ", lda=%" PRId64 ", ldc=%" PRId64 ", "
            "order=%c, uplo=%c, trans=%c\n", (int64_t)N, (int64_t)K, (int64_t)lda, (int64_t)ldc,
            order_to_char(order), uplo_to_char(uplo), transpose_to_char(trans));
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    nvpl_int_t rowsA = (trans == CblasNoTrans) ? N : K;
    nvpl_int_t colsA = (trans == CblasNoTrans) ? K : N;
    nvpl_int_t rowsC = N;
    nvpl_int_t colsC = N;

    // allocate memory
    A = (nvpl_dcomplex_t *)malloc(lda * colsA  * sizeof(nvpl_dcomplex_t));
    C = (nvpl_dcomplex_t *)malloc(ldc * colsC  * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zmatrix(A, rowsA, colsA, lda, order, Full, CblasNonUnit);
    fill_zmatrix(C, rowsC, colsC, ldc, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);

    // print input data
    print_zmatrix(A, rowsA, colsA, lda, "A", order);
    print_zmatrix(C, rowsC, colsC, ldc, "C", order);

    // call cblas_zherk
    cblas_zherk(order, uplo, trans, N, K, alpha, A, lda, beta, C, ldc);

    // print result
    printf("\nAfter cblas_zherk operation:\n");
    print_zmatrix(C, rowsC, colsC, ldc, "C", order);

    // release memory
    free(A);
    free(C);
    return 0;
}

File: nvpl_blas/f77/zherk.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZHERK 
!
!******************************************************************************
      program   ZHERK_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1      uplo, trans
      parameter        (uplo='U', trans='N')
      integer          n, k, i, j
      integer          lda, ldc
      parameter        (n=2, k=2,
     &                  lda=2, ldc=2)
      double precision alpha, beta
      parameter        (alpha=(1.0, 2.0), beta=(1.0, 2.0))
      double complex   a(lda,k), c(ldc,n)

*       External Subroutines
      external        print_zmatrix, fill_zmatrix
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, k, lda, ldc, uplo, trans
      print 101, alpha, beta
      call fill_zmatrix(a,n,k,lda,'F','N')
      call fill_zmatrix(c,n,n,ldc,uplo,'N')
      call print_zmatrix(a,n,k,lda,'A')
      call print_zmatrix(c,n,n,ldc,'C')
*
*      Call ZHERK subroutine
      call ZHERK(uplo,trans,n,k,alpha,a,lda,beta,c,ldc)
*
      print*
      print 102, 'After ZHERK operation:'
      call print_zmatrix(c,n,n,ldc,'C')

      stop
 99   format('Example: ZHERK for the Hermitian rank-k update')
 100  format('#### args: n=',i1,', k=',i1,', lda=',i1,
     &       ', ldc=',i1,', uplo=',a1,', trans=',a1)
 101  format(11x,'alpha=',f3.1,'  beta=',f3.1)
 102  format(A)
      end
